crates/rspack_core/src/external_module.rs

```diff
@@ -150,6 +151,31 @@ impl ExternalModule {
     user_request: String,
     dependency_meta: DependencyMeta,
   ) -> Self {
+    let mut resolved_external_type: ExternalType = external_type.clone();
+    if external_type == "module-import" {
+      if let Some(r#type) = &dependency_meta.external_type {
+        resolved_external_type = match r#type {
+          ExternalTypeEnum::Import => "import".to_string(),
+          ExternalTypeEnum::Module => "module".to_string(),
+        };
+      } else {
+        let presets = &dependency_meta.externals_presets;
+        if presets.web {
+          resolved_external_type = "module".to_string();
+        } else if presets.web_async {
+          resolved_external_type = "import".to_string();
+        } else if presets.node
```

Comment:
externals_presets should not effect the external_type here, externalsPresets is just a "preset" of externals, it should be resolved at config-level, instead of the core logic

Reply:
As reference of https://github.com/webpack/webpack/pull/18620#issuecomment-2276069970.

> 1. Work with `externalsPresets`.
>    - when `externalsPresets` is a node.js-related presets, fallback to `node-commonjs`;
>    - when `externalsPresets === 'web'`, fallback to `module`;
>    - when `externalsPresets === 'web-async'`, fallback to `import`;
> 
>    If `externalsPresets` is not specified, we could set the fallback to `commonjs` (🤔 I'm not sure, this might break `module-import`'s semantics)
> 1. Adding `module-import-node-commonjs`, `module-import-import`, `module-import-module`, `module-import-commonjs`, maybe too verbose.
> 3. Adding a new config `exteralsFallback`, but this will only work with `"module-import"`, but adding an option on the top is a high tradeoff.

Plan 2 or 3 is better at decoupling the configs. Which one do u think is better (we still have chance to change the implementation before webpack's next release), or other better ideas?

Reply:
https://github.com/webpack/webpack/blob/main/lib/WebpackOptionsApply.js#L143-L155

Try using a function for dependency type awareness external type